### PR TITLE
support interval.microseconds for range window TimeStampType

### PR DESCRIPTION
**Pull request by wbo4958**
1. This change supports interval.microseconds for TimeStampType.

2.fix a range window on DateType bug.

  When the order-by column type is DateType, the value bounded type
  should be IntegerType instead of CalendarIntervalType.
  In other words, Interval is not working when order-by column is DateType.

Signed-off-by: Bobby Wang <wbo4958@gmail.com>

### CODE REVIEW

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 1]
```diff
@@ -600,13 +601,12 @@ class GpuSpecifiedWindowFrameMeta(
...
-            if (interval.microseconds != 0 || interval.months != 0) { // DAYS == 0 is permitted.
-              willNotWorkOnGpu(s"Bounds for Range-based window frames must be specified" +
-                s" only in DAYS. Found $interval")
+          case Literal(ci: CalendarInterval, CalendarIntervalType) =>
+            if (ci.months != 0) {
+              willNotWorkOnGpu("interval months isn't supported")
             }
-            interval.days.toLong
+            // return the total microseconds
+            ci.days * TimeUnit.DAYS.toMicros(1) + ci.microseconds
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:609]: This calculation can overflow. In the normal SQL case it should be very unlikely to happen, but we should still detect it before it happens and fall back to the CPU.

Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala]: Yeah, the overflow checking is quite necessary, Thx for the commenting. Done

**Comment by wbo4958** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala [hunk 2]
```diff
@@ -600,16 +601,23 @@ class GpuSpecifiedWindowFrameMeta(
...
+              willNotWorkOnGpu("interval months isn't supported")
+            }
+            // return the total microseconds
+            try {
+              Math.addExact(
+                Math.multiplyExact(ci.days.toLong, TimeUnit.DAYS.toMicros(1)),
+                ci.microseconds)
+            } catch {
+              case e: ArithmeticException =>
+                willNotWorkOnGpu("interval total microseconds has overflow")
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala:615]: I would like to see a better message. It is not clear how the user can fix this or what exactly has happened.

Perhaps something more like

```
willNotWorkOnGpu(s"windows over timestamps are converted to microseconds and $ci is too large to fit")
```

**Review by revans2 — COMMENTED** [review]
Just a nit

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala (hunk 2 above)
Comment (wbo4958) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala]: Thx @revans2. Done

**Comment by wbo4958** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**